### PR TITLE
plugin download: also invalidate GITHUB_TOKEN on 403

### PR DESCRIPTION
**Pull request by tgummerer**
Similar to what we did in https://github.com/pulumi/pulumi/pull/17351, we should also invalidate the GITHUB_TOKEN when the GitHub API returns a 403, which can happen in some cases when the token isn't valid.

Fixes https://github.com/pulumi/pulumi/issues/17217

### CODE REVIEW

File: sdk/go/common/workspace/plugins.go [hunk 1]
```diff
@@ -447,24 +447,21 @@ func (source *githubSource) getHTTPResponse(
...
+	// Wrap 403 rate limit errors with a more helpful message later. This is a rate limiting error only
+	// if x-ratelimit-remaining is 0.
+	// https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#exceeding-the-rate-limit
 	var downErr *downloadError
-	if !errors.As(err, &downErr) || downErr.code != 403 {
-		// If we see a 401 error and were using a token we'll disable that token and try again
-		if downErr != nil && downErr.code == 401 && source.token != "" {
+	if !errors.As(err, &downErr) || downErr.code != 403 || downErr.header.Get("x-ratelimit-remaining") != "0" {
+		// If we see a 401 error and we're using a token we'll disable that token and try again
+		fmt.Println(err, downErr)
```
Comment (julienp) [inline sdk/go/common/workspace/plugins.go:456]:
```suggestion
```

File: sdk/go/common/workspace/plugins.go [hunk 2]
```diff
@@ -447,24 +447,21 @@ func (source *githubSource) getHTTPResponse(
...
+		if downErr != nil && (downErr.code == 401 || downErr.code == 403) && source.token != "" {
```
Comment (julienp) [inline sdk/go/common/workspace/plugins.go:457]: I have a hard time with the nested if's and figuring out all the paths through here. I think this is right, but maybe we should split this out or comment differently? The comment on line 450 says we're wrapping 403 errors, why are we *not* wrapping a 403 sometimes here?

File: sdk/go/common/workspace/plugins_test.go [hunk 3]
```diff
@@ -740,6 +740,63 @@ func TestPluginDownload(t *testing.T) {
...
+			Version:           &version,
+			Kind:              apitype.PluginKind("resource"),
+		}
+		source, err := spec.GetSource()
+		require.NoError(t, err)
+		attempts := 0
+		getHTTPResponse := func(req *http.Request) (io.ReadCloser, int64, error) {
+			attempts++
+
+			fmt.Println("req.Header.Get(\"Authorization\")", req.Header.Get("Authorization"))
```
Comment (julienp) [inline sdk/go/common/workspace/plugins_test.go:758]: There's a print here and below, are these intentionally left in?

Comment (tgummerer) [reply sdk/go/common/workspace/plugins_test.go:758]: Doh, no I was just debugging and forgot these.

File: sdk/go/common/workspace/plugins.go (hunk 2 above)
Comment (tgummerer) [reply sdk/go/common/workspace/plugins.go:463]: I tried improving the comment here, and split out a function naming rate limit errors specifically.  What do you think?

Comment (julienp) [reply sdk/go/common/workspace/plugins.go:463]: This is great, thanks!

**Review by julienp — APPROVED** [review]

**Pull request merged**